src/docs/binding-generics.md

```diff
@@ -236,3 +236,194 @@ The proposed API will only work with types known at compile time. In some cases
+
+The frozen keyword can be applied to a generic struct just like it can be applied to a regular struct. However because the generic parameters type (and hence size and layout) is unknown at the compile time, so is the size and the layout of the generic struct. Specifically, the generic parameter can be a frozen or a non-frozen struct.
+
+If we consider the four cases:
+
+1. Non-frozen struct and at least one non-frozen generic parameter
+2. Non-frozen struct and (recursively) frozen generic parameters
+3. Frozen struct and at least one non-frozen generic parameter
+4. Frozen struct and (recursively) frozen generic parameters
+
+The cases one to three are simple from the projection perspective. The instantiated generic will be passed using indirection (as a non-frozen struct would in previous examples).
+
+The case four is problematic, as instances of a closed generic can be passed using swift lowering algorithm. Lets look into this more detail:
+
+```swift
+@frozen
+public struct Pair<T, U> {...}
+
+@frozen
+public struct FrozenStruct {
+    let a: Double
+    let b : Double
+}
+
+public struct NonFrozenStruct {
+    let a: Double
+    let b : Double
+}
+
+// Case 3
+public func withAtLeastOneNonFrozenParameter(data: Pair<FrozenStruct, NonFrozenStruct>) -> Pair<FrozenStruct, NonFrozenStruct> {
+    return data;
+
+// Case 4
+public func withTwoFrozenParameters(data: Pair<FrozenStruct, FrozenStruct>) -> Pair<FrozenStruct, FrozenStruct> {
+    return data;
+}
+}
+```
+
+The corresponding llvm-ir signatures of the above two functions are as follows (cases one and two would generate a signature very similar to the signature of case three).
+
+```llvm
+// Case 3
+void @"output.withAtLeastOneNonFrozenParameter(data: output.Pair<output.FrozenStruct, output.NonFrozenStruct>) -> output.Pair<output.FrozenStruct, output.NonFrozenStruct>"(ptr noalias nocapture sret(%T6output13PairVyAA0D6StructVAA0cdE0VG) %0, ptr noalias nocapture dereferenceable(32) %1)
+
+// Case 4
+{ double, double, double, double } @"output.withTwoFrozenParameters(data: output.Pair<output.FrozenStruct, output.FrozenStruct>) -> output.Pair<output.FrozenStruct, output.FrozenStruct>"(double %0, double %1, double %2, double %3)
+```
+
+Case four passes the Pair using lowering algorithm.
+
+From the user perspective we would like to have a function with the following signature:
+
+```csharp
+    public static Pair<FrozenStruct, FrozenStruct> WithTwoFrozenParameters(Pair<FrozenStruct, FrozenStruct> pair)
+```
+
+Which would then ideally call the following PInvoke.
+
+```csharp
+    [UnmanagedCallConv(CallConvs = [typeof(CallConvSwift)])]
+    public static extern Pair<FrozenStruct, FrozenStruct> WithTwoFrozenParameters(Pair<FrozenStruct, FrozenStruct> pair);
+```
+
+This however would not work as `Pair<FrozenStruct, FrozenStruct>` is a class. Runtime would try to pass the value by reference (it does not even know the layout to begin with). We could solve this problem on the projection layer.
```

Comment:
I think the problem is that user-facing C# `Pair` is a class. In the specific case of `Pair<FrozenStruct, FrozenStruct>`, it could be a struct (and that's the wrapper introduced later) but we would have to have 2x C# `Pair` to handle the case of `NonFrozenStruct` as well. @jkurdek please correct me if I'm wrong here.

Reply:
>If a frozen struct doesn't require a copy constructor when moved in memory, I would still consider it a value type (struct). 

So for generics strictly. I think we cannot tell at compile time whether a generic will have or wont have to use a copy constructor as we have no idea what the generic parameter will be. Some instances of the generic, `Pair<SomethingThatRequiresCopyConstructor, int>` will require it. Some like `Pair< FrozenStruct, FrozenStruct>` will not. 

@matouskozak I think you got it right. `Pair<FrozenStruct, FrozenStruct>` would ideally be represented as struct. 

>Additionally, as we discussed offline, the projection will flip if a generic used becomes frozen. From the user’s perspective, this can occur without their control, which is my primary concern when introducing this rule.

 Making something non-frozen frozen is an ABI breaking change. I do not think we can do much about it. One way of solving this would be representing every struct as a class on C# side and abstracting the implementation details

Reply:
> One way of solving this would be representing every struct as a class on C# side and abstracting the implementation details

Do you mean every generic struct? In this case, the size and alignment of an object of opaque layout, as well as whether it is trivial or bitwise movable, is determined by querying its value witness table. If contains frozen structs, the helper can be used.

Reply:
I meant all structs. Its more of stretch idea thing than actual proposal. I feel that projecting some as structs and some as classes creates a weird UX for developers who are not Swift professionals. I think would prefer to have uniformity - e.g. all structs to be classes. I do understand however that this would require a lot custom runtime work and would bring its own set of problems. Anyway it is a different discussion.

Reply:
> I meant all structs. Its more of stretch idea thing than actual proposal. I feel that projecting some as structs and some as classes creates a weird UX for developers who are not Swift professionals.

It seems we need this layer of abstraction for structs with opaque layouts (those not known until runtime). In other cases, this abstraction isn’t necessary. We should prioritize simplicity, ensuring that users reading Swift docs know what to expect. If a struct can be projected directly, we should do so.

> I do understand however that this would require a lot custom runtime work and would bring its own set of problems. Anyway it is a different discussion.

Let's do cost estimate for this.

/cc: @jkoritzinsky

Reply:
Since we moved frozen type with frozen generic parameters to https://github.com/dotnet/runtimelab/issues/2797, let's resolve this comment. Other cases should be covered with the helper marshalling struct.

Reply:
Let's make sure we see the same picture before resolving. How I view it is that the marshalling struct is brought specifically to solve the frozen type with frozen generic problem. non-frozen generic or non-frozen parameters will work out of the box. So in my eyes the #2797 issue doesn't really concern generics but rather struct / class projection of non-generic structs.